feat: add GraphQL to fetch notifications

PrinceYadav2 opened the pull request:
**Description**:
This PR adds graphql and to fetch all the notifications

**Related changes:**
Implement graphql for application root
Add a function to fetch all the notifications
Add Notification Resolver
Register Returning Models by modifying notification entity.

**Curl:** 
```curl

curl 'http://localhost:3000/graphql' -H 'Accept-Encoding: gzip, deflate, br' -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Connection: keep-alive' -H 'DNT: 1' -H 'Origin: http://localhost:3000' -H 'Authorization: Bearer <SERVER_AUTH_TOKEN>' --data-binary '{"query":"query {\n  notifications {\n    channelType\n    createdBy\n    createdOn\n    data\n    deliveryStatus\n    id\n    result\n    status\n    updatedBy\n  }\n}"}' --compressed
```


**Sample Response:**
```
{
  "data": {
    "notifications": [
      {
        "channelType": 2,
        "createdBy": "osmo_notify",
        "createdOn": "2023-11-28T04:59:32.000Z",
        "data": {
          "from": "yadav.prince003@gmail.com",
          "to": "prince.y@osmosys.co",
          "subject": "Test subject",
          "text": "This is a test notification",
          "html": "<b>This is a test notification</b>"
        },
        "deliveryStatus": 3,
        "id": 1,
        "result": {
          "result": {
            "status": 200,
            "id": "<20231128045933.88c9fb88ce0aee6a@mg.pinestem.com>",
            "message": "Queued. Thank you."
          }
        },
        "status": 1,
        "updatedBy": "osmo_notify"
      },
      {
        "channelType": 1,
        "createdBy": "osmo_notify",
        "createdOn": "2023-11-28T05:04:30.000Z",
        "data": {
          "from": "yadav.prince003@gmail.com",
          "to": "prince.y@osmosys.co",
          "subject": "Test subject",
          "text": "This is a test notification",
          "html": "<b>This is a test notification</b>"
        },
        "deliveryStatus": 4,
        "id": 2,
        "result": {
          "result": {
            "errno": -3008,
            "code": "EDNS",
            "syscall": "getaddrinfo",
            "hostname": "smtp.gmail.com",
            "command": "CONN"
          }
        },
        "status": 1,
        "updatedBy": "osmo_notify"
      }
    ]
  }
}
```

**Snapshot:**
Success:
![image](https://github.com/OsmosysSoftware/osmo-notify/assets/114974207/87cd955d-7213-41bf-a7b5-24ca6815b9e0)
Fail:
![image](https://github.com/OsmosysSoftware/osmo-notify/assets/114974207/e64d92d3-8ee4-41fd-b15b-2ac985082513)

xixas commented on src/modules/notifications/notifications.resolver.ts:
```diff
@@ -0,0 +1,13 @@
+
+  @Query(() => [Notification], { name: 'notifications' })
+  async findAll(@Context() context) {
+    return this.notificationsService.getAllNotifications();
```
Get notification whose status is 1.

PrinceYadav2 replied:
Fixed in the commit: https://github.com/OsmosysSoftware/osmo-notify/pull/68/commits/07cbc5def15a46d0b95364563c2c2c035efbfe18

xixas commented on src/modules/notifications/notifications.resolver.ts:
```diff
@@ -0,0 +1,13 @@
+    constructor(private readonly notificationsService: NotificationsService) {}
+
+  @Query(() => [Notification], { name: 'notifications' })
+  async findAll(@Context() context) {
```
Add proper example in description I see **# Write your query or mutation here**

xixas replied:
Fix lint error

PrinceYadav2 replied:
Fixed in the commit https://github.com/OsmosysSoftware/osmo-notify/pull/68/commits/07cbc5def15a46d0b95364563c2c2c035efbfe18

PrinceYadav2 replied:
> Add proper example in description I see **# Write your query or mutation here**

My bad, that was a comment that got exported in curl, Removed it.
Added a screenshot in description as well.

xixas commented:
@PrinceYadav2 
We have recently also added API key guard.
Add this for graphQL too
https://github.com/OsmosysSoftware/osmo-notify/blob/7778edb226a7fa83e0bf0fa360a4760278fc8bde/src/modules/notifications/notifications.controller.ts#L26

PrinceYadav2 commented:
> @PrinceYadav2 We have recently also added API key guard. Add this for graphQL too
> 
> https://github.com/OsmosysSoftware/osmo-notify/blob/7778edb226a7fa83e0bf0fa360a4760278fc8bde/src/modules/notifications/notifications.controller.ts#L26

Added API key guard for the same, attached a screenshot of failing case in description.

xixas commented on apps/api/src/common/guards/api-key/api-key.guard.ts:
```diff
@@ -9,7 +10,19 @@ export class ApiKeyGuard implements CanActivate {
     const request = context.switchToHttp().getRequest();
 
-    const authHeader = request.headers['authorization'];
+    // Get auth header incase of http request
```
Did you see if this works for the REST API too(POST request in controller)?

PrinceYadav2 replied:
Yes, @vikas-kyatannawar ,
I verified both the cases for rest request as well, It is working fine for it.

![image](https://github.com/OsmosysSoftware/osmo-notify/assets/114974207/3e3ce710-86cb-41a1-9042-dd1ec958e7ce)
